File: server/tests/test_serializer.cpp
#include <cstdint>
#include <memory>
#include <string>
#include <vector>

#include <gtest/gtest.h>
#include <nlohmann/json.hpp>

#include "adapters/json_serializer.h"
#include "ports/i_serializer.h"

using iris::ISerializer;
using iris::JsonSerializer;
using nlohmann::json;

class SerializerTest : public ::testing::Test {
protected:
    std::unique_ptr<ISerializer> serializer =
        std::make_unique<JsonSerializer>();
};

TEST_F(SerializerTest, RoundTrip_SimpleObject) {
    json obj = {{"cmd", "add"}, {"args", {10, 32}}};
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

TEST_F(SerializerTest, RoundTrip_Integer) {
    json obj = 42;
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

TEST_F(SerializerTest, RoundTrip_String) {
    json obj = "hello world";
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

TEST_F(SerializerTest, RoundTrip_Array) {
    json obj = {1, "two", 3.0, true, nullptr};
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

// ---------- nested objects ----------

TEST_F(SerializerTest, RoundTrip_NestedObject) {
    json obj = {
        {"id", "abc-123"},
        {"status", "ok"},
        {"result", {
            {"values", {1, 2, 3}},
            {"metadata", {{"source", "test"}}}
        }}
    };
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

// ---------- unicode ----------

TEST_F(SerializerTest, RoundTrip_Unicode) {
    json obj = {
        {"greeting", u8"こんにちは"},
        {"emoji", u8"🚀🌍"},
        {"accent", u8"café résumé"}
    };
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_EQ(result, obj);
}

// ---------- empty structures ----------

TEST_F(SerializerTest, RoundTrip_EmptyObject) {
    json obj = json::object();
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_TRUE(result.is_object());
    EXPECT_TRUE(result.empty());
}

TEST_F(SerializerTest, RoundTrip_EmptyArray) {
    json obj = json::array();
    auto bytes = serializer->serialize(obj);
    auto result = serializer->deserialize(bytes);
    EXPECT_TRUE(result.is_array());
    EXPECT_TRUE(result.empty());
}

// ---------- error case ----------

TEST_F(SerializerTest, DeserializeInvalidJson) {
    std::vector<uint8_t> bad = {'n', 'o', 't', 'j', 's', 'o', 'n'};
    EXPECT_THROW(serializer->deserialize(bad), std::runtime_error);
}

File: client/tests/test_serializer.py
"""Unit tests for the JsonSerializer adapter."""

import json

import pytest

from src.adapters.json_serializer import JsonSerializer


@pytest.fixture
def serializer() -> JsonSerializer:
    return JsonSerializer()


# ---------- round-trip ----------


def test_round_trip_simple_object(serializer: JsonSerializer) -> None:
    obj = {"cmd": "add", "args": [10, 32]}
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


def test_round_trip_integer(serializer: JsonSerializer) -> None:
    obj = 42
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


def test_round_trip_string(serializer: JsonSerializer) -> None:
    obj = "hello world"
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


def test_round_trip_array(serializer: JsonSerializer) -> None:
    obj = [1, "two", 3.0, True, None]
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


# ---------- nested objects ----------


def test_round_trip_nested_object(serializer: JsonSerializer) -> None:
    obj = {
        "id": "abc-123",
        "status": "ok",
        "result": {
            "values": [1, 2, 3],
            "metadata": {"source": "test"},
        },
    }
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


# ---------- unicode ----------


def test_round_trip_unicode(serializer: JsonSerializer) -> None:
    obj = {
        "greeting": "こんにちは",
        "emoji": "🚀🌍",
        "accent": "café résumé",
    }
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj


# ---------- empty structures ----------


def test_round_trip_empty_object(serializer: JsonSerializer) -> None:
    obj: dict = {}
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj
    assert isinstance(result, dict)


def test_round_trip_empty_array(serializer: JsonSerializer) -> None:
    obj: list = []
    result = serializer.deserialize(serializer.serialize(obj))
    assert result == obj
    assert isinstance(result, list)


# ---------- error case ----------


def test_deserialize_invalid_json(serializer: JsonSerializer) -> None:
    with pytest.raises(json.JSONDecodeError):
        serializer.deserialize(b"not json")
